File: frontend/src/lib/demo/optimiser.ts
/**
 * Pack the cards worth grading into submissions, ported from the backend.
 *
 * Three passes, and the third is the one that matters:
 *
 * 1. **Route** every analysable card at a batch big enough for the bulk tiers
 *    to be on the table, and take the route the engine recommends.
 * 2. **Pack** by (company, tier), checking each tier's own minimum.
 * 3. **Re-verify** every card at the size its batch actually came out at.
 *
 * Without step 3 the plan looks profitable and is not: a card that clears the
 * bar at twenty-five may not at six, because six cards carry the postage
 * twenty-five would have shared.
 */

import type { CardEvaluation, GradingCompany, OptimiserPlan, PlacedCard, ProposedBatch } from '@/lib/types'
import { toMinor } from './market'

/** Decisions that mean the card belongs in a submission at all. */
const WORTH_GRADING = new Set(['grade', 'grade_if_batch_filled'])

const money = (minor: number | null | undefined) =>
  minor === null || minor === undefined
    ? '—'
    : `${minor < 0 ? '−' : ''}£${(Math.abs(minor) / 100).toFixed(2)}`

export interface Candidate {
  cardId: string
}

/**
 * A batch big enough that every tier is on the table. Routing at 1 would hide
 * the bulk tiers and send everything to the expensive ones.
 */
export function hopefulBatchSize(companies: GradingCompany[]): number {
  const minimums = companies
    .flatMap((company) => company.tiers)
    .filter((tier) => tier.active)
    .map((tier) => tier.minimum_cards)
  return Math.max(1, ...minimums)
}

export function optimise(input: {
  candidates: string[]
  totalCards: number
  companies: GradingCompany[]
  currency: string
  limit: number
  /** Evaluate one card at one batch size. */
  evaluate: (cardId: string, batchSize: number) => CardEvaluation
}): OptimiserPlan {
  const { companies, currency, evaluate } = input
  const plan: OptimiserPlan = {
    status: 'ok',
    reason: null,
    currency,
    analysable: input.candidates.length,
    worth_grading: 0,
    placed: 0,
    total_cards: input.totalCards,
    truncated: false,
    routed_at_batch_size: hopefulBatchSize(companies),
    expected_profit: null,
    total_grading_cost: null,
    batches: [],
    unplaced: [],
    stopped_paying: [],
    notes: [],
  }

  let candidates = input.candidates
  if (candidates.length > input.limit) {
    plan.truncated = true
    candidates = candidates.slice(0, input.limit)
  }

  // --- Pass 1: route ------------------------------------------------------
  const routed = new Map<string, { evaluation: CardEvaluation; cardId: string }[]>()
  for (const cardId of candidates) {
    const evaluation = evaluate(cardId, plan.routed_at_batch_size)
    const recommendation = evaluation.recommendation
    if (!WORTH_GRADING.has(recommendation.decision)) continue
    plan.worth_grading += 1

    if (!recommendation.company_code) {
      plan.unplaced.push({
        card_id: cardId,
        name: evaluation.raw.display_name,
        set_label: evaluation.raw.set_label,
        company_code: null,
        tier_name: null,
        expected_profit: recommendation.expected_profit,
        reason: 'The engine recommends grading it but named no grader to send it to.',
      })
      continue
    }
    const key = `${recommendation.company_code}::${recommendation.tier_name ?? ''}`
    const members = routed.get(key) ?? []
    members.push({ evaluation, cardId })
    routed.set(key, members)
  }

  if (!routed.size) {
    plan.status = 'insufficient_data'
    plan.reason = plan.analysable
      ? 'Nothing in your collection clears the bar for grading right now, so there is no ' +
        'submission to build.'
      : 'No card has both a condition assessment and comparable sales yet, so nothing can be ' +
        'decided — assess a card and add its sales.'
    finish(plan, input.limit)
    return plan
  }

  // --- Passes 2 and 3: pack, then re-verify at the real size ---------------
  const ordered = [...routed.entries()].sort((a, b) => b[1].length - a[1].length)
  for (const [key, members] of ordered) {
    const [companyCode, tierName] = key.split('::')
    const company = companies.find((item) => item.code === companyCode)
    if (!company) continue
    const tier = company.tiers.find((item) => item.tier_name === tierName && item.active) ?? null

    const realSize = members.length
    const batch: ProposedBatch = {
      company_id: company.id,
      company_code: companyCode,
      tier_id: tier?.id ?? null,
      tier_name: tierName || null,
      effective_tier_name: tierName || null,
      card_count: realSize,
      minimum_cards: tier?.minimum_cards ?? 1,
      maximum_cards: tier?.maximum_cards ?? null,
      short_by: 0,
      expected_profit: null,
      grading_cost: null,
      expected_profit_if_filled: null,
      viable: true,
      reason: null,
      cards: [],
    }

    const short = tier !== null && realSize < tier.minimum_cards
    if (short && tier) {
      batch.short_by = tier.minimum_cards - realSize
      batch.viable = false
    }

    let profitMinor = 0
    let costMinor = 0
    for (const { evaluation, cardId } of members) {
      const placed = reverify(evaluate, cardId, evaluation, batch, realSize)
      batch.cards.push(placed)
      if (placed.still_pays) {
        profitMinor += toMinor(placed.expected_profit ?? 0)
        costMinor += toMinor(placed.grading_cost ?? 0)
      } else {
        plan.stopped_paying.push(placed)
      }
    }

    const landed = new Set(
      batch.cards.filter((card) => card.still_pays).map((card) => card.tier_name),
    )
    batch.effective_tier_name = landed.size === 1 ? [...landed][0] : (tierName || null)

    if (short && tier) {
      const filled = members.reduce(
        (sum, { evaluation }) => sum + toMinor(evaluation.recommendation.expected_profit ?? 0),
        0,
      )
      batch.expected_profit_if_filled = Number((filled / 100).toFixed(2))
      const sameTier = batch.effective_tier_name === tierName
      batch.reason =
        `${companyCode} ${tierName} needs ${tier.minimum_cards} cards and this batch has ` +
        `${realSize}. ` +
        (sameTier ? '' : `As it stands these would be graded at ${batch.effective_tier_name}. `) +
        `Adding ${batch.short_by} more card(s) at this tier is worth ` +
        `${money(filled - profitMinor)} across the cards already in it.`
    }

    batch.expected_profit = Number((profitMinor / 100).toFixed(2))
    batch.grading_cost = Number((costMinor / 100).toFixed(2))
    plan.batches.push(batch)
  }

  plan.batches.sort((a, b) => (b.expected_profit ?? 0) - (a.expected_profit ?? 0))
  const paying = plan.batches.flatMap((batch) => batch.cards.filter((card) => card.still_pays))
  plan.placed = paying.length
  if (paying.length) {
    plan.expected_profit = Number(
      (paying.reduce((sum, card) => sum + toMinor(card.expected_profit ?? 0), 0) / 100).toFixed(2),
    )
    plan.total_grading_cost = Number(
      (paying.reduce((sum, card) => sum + toMinor(card.grading_cost ?? 0), 0) / 100).toFixed(2),
    )
  }

  finish(plan, input.limit)
  return plan
}

/**
 * Cost this card again at the size its batch actually came out at.
 *
 * `grade_if_batch_filled` was a fair answer while the batch was hypothetical.
 * Here the size is a known fact, so it means "not in the batch you have".
 */
function reverify(
  evaluate: (cardId: string, batchSize: number) => CardEvaluation,
  cardId: string,
  routedEval: CardEvaluation,
  batch: ProposedBatch,
  realSize: number,
): PlacedCard {
  const routed = routedEval.recommendation
  const placed: PlacedCard = {
    card_id: cardId,
    name: routedEval.raw.display_name,
    set_label: routedEval.raw.set_label,
    company_code: batch.company_code,
    tier_id: batch.tier_id,
    tier_name: batch.tier_name,
    declared_value: routedEval.grading_options.declared_value,
    decision_when_routed: routed.decision,
    decision_in_batch: routed.decision,
    expected_profit: routed.expected_profit,
    grading_cost: routed.grading_cost,
    opportunity_score: routed.opportunity_score,
    still_pays: true,
    reason: null,
    cheaper_tier_name: null,
    cheaper_tier_saving: null,
  }

  if (realSize === routed.assumed_batch_size) {
    suggestCheaperTier(placed, routedEval)
    return placed
  }

  const actual = evaluate(cardId, realSize)
  const recommendation = actual.recommendation
  placed.decision_in_batch = recommendation.decision
  placed.opportunity_score = recommendation.opportunity_score

  // The tier the card actually lands on at this count, which is not always the
  // one it was routed to: Bulk needs twenty-five, so in a batch of nine these
  // are Economy cards. Reporting them as Bulk at Economy's price would quote a
  // route that does not exist at this size.
  placed.tier_name = recommendation.tier_name ?? batch.tier_name
  placed.tier_id = placed.tier_name === batch.tier_name ? batch.tier_id : null

  const atThisSize = actual.expected_outcomes.outcomes.find(
    (outcome) =>
      outcome.company_code === batch.company_code && outcome.tier_name === placed.tier_name,
  )
  placed.expected_profit = atThisSize?.expected_profit ?? recommendation.expected_profit
  placed.grading_cost = atThisSize?.grading_cost ?? recommendation.grading_cost

  if (recommendation.decision !== 'grade') {
    placed.still_pays = false
    const was = routed.expected_profit
    const now = placed.expected_profit
    const moved =
      was !== null && now !== null && toMinor(was) !== toMinor(now)
        ? ` Expected profit falls from ${money(toMinor(was))} to ${money(toMinor(now))}.`
        : ''
    placed.reason =
      `Worth grading in a batch of ${routed.assumed_batch_size}, but this batch has ` +
      `${realSize} — so it carries a bigger share of the postage.${moved}`
  } else {
    suggestCheaperTier(placed, actual)
  }
  return placed
}

/** "Move it to Bulk and save £2.20" — but only where the move is legal. */
function suggestCheaperTier(placed: PlacedCard, evaluation: CardEvaluation): void {
  const options = evaluation.grading_options.options
  const current = options.find(
    (option) =>
      option.company_code === placed.company_code && option.tier_name === placed.tier_name,
  )
  if (!current || current.total_cost === null) return

  const cheaper = options.filter(
    (option) =>
      option.company_code === placed.company_code &&
      option.available &&
      option.total_cost !== null &&
      option.total_cost < current.total_cost!,
  )
  if (!cheaper.length) return

  const best = cheaper.reduce((low, option) =>
    (option.total_cost ?? 0) < (low.total_cost ?? 0) ? option : low,
  )
  placed.cheaper_tier_name = best.tier_name
  placed.cheaper_tier_saving = Number((current.total_cost - (best.total_cost ?? 0)).toFixed(2))
}

function finish(plan: OptimiserPlan, limit: number): void {
  if (plan.truncated) {
    plan.notes.push(
      `Only the ${limit} most recently updated ready cards were considered. Raise the limit ` +
        'to include the rest.',
    )
  }
  if (plan.stopped_paying.length) {
    plan.notes.push(
      `${plan.stopped_paying.length} card(s) were worth grading in a full batch but not in the ` +
        'batch they ended up in. They are listed against their batch, and are not counted in ' +
        'any total.',
    )
  }
  const short = plan.batches.filter((batch) => !batch.viable)
  if (short.length) {
    plan.notes.push(
      `${short.length} proposed batch(es) are short of their tier's minimum. They are costed ` +
        'as they stand, so you can see what filling them would be worth.',
    )
  }

  if (plan.status === 'insufficient_data') return
  if (plan.truncated || short.length || plan.stopped_paying.length || plan.unplaced.length) {
    plan.status = 'partial'
    plan.reason = plan.notes[0] ?? null
  } else {
    plan.status = 'ok'
  }
}

File: backend/app/services/optimiser.py
"""Pack the cards worth grading into submissions that are actually sendable.

The decision engine answers *whether* a card is worth grading. This answers
*how to send it* — and those two questions are entangled in a way that is easy
to get quietly wrong.

Grading costs depend on the batch. A card that clears your bar at twenty-five
may not clear it at six, because six cards carry the postage twenty-five would
have shared. So an optimiser that groups cards by "what the engine recommended"
and stops there is proposing submissions it has never actually costed.

This one does three passes:

1. **Route.** Evaluate every analysable card at a batch big enough for the bulk
   tiers to be on the table, and take the route the engine recommends. That is
   the card's opinion of where it wants to go *if* a batch exists.

2. **Pack.** Group by (company, tier). Each group is a candidate submission, and
   each is checked against that tier's own minimum — which is a minimum at that
   tier, not in the parcel.

3. **Re-verify.** Cost every card again at the size its batch actually came out
   at. Cards that no longer pay are reported, with the reason and the number
   that changed. They are never silently shipped and never silently dropped.

Step 3 is the one that matters. Without it the optimiser produces a plan that
looks profitable and is not, which is precisely the failure the whole
application exists to avoid.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.enums import BlockStatus, Decision
from app.models import Card, ConditionAssessment, GradingCompany, GradingTier, MarketPrice
from app.money import format_money, to_minor
from app.services import evaluation, settings_service

__all__ = ["OptimiserResult", "PlacedCard", "ProposedBatch", "optimise"]

#: Evaluating a card costs about 20ms and the optimiser evaluates each one
#: twice — once to route it, once to re-verify it at its real batch size. Above
#: this many candidates the sweep is cut, and the cut is always reported.
DEFAULT_LIMIT = 150

#: Decisions that mean the card belongs in a submission at all.
_WORTH_GRADING = {Decision.GRADE.value, Decision.GRADE_IF_BATCH_FILLED.value}


@dataclass
class PlacedCard:
    """One card in a proposed batch, costed at that batch's real size."""

    card_id: str
    name: str
    set_label: str | None = None
    company_code: str | None = None
    tier_id: str | None = None
    tier_name: str | None = None
    declared_value: float | None = None

    #: What the engine said when the card was routed, at the hopeful batch size.
    decision_when_routed: str = Decision.INSUFFICIENT_DATA.value
    #: What it says at the size this batch actually came out at.
    decision_in_batch: str = Decision.INSUFFICIENT_DATA.value
    expected_profit: float | None = None
    grading_cost: float | None = None
    opportunity_score: float | None = None

    #: False when the card stopped paying once the real batch size was known.
    still_pays: bool = True
    #: Why it stopped, in the user's terms.
    reason: str | None = None
    #: A cheaper tier for this card in this same batch, when one exists.
    cheaper_tier_name: str | None = None
    cheaper_tier_saving: float | None = None


@dataclass
class ProposedBatch:
    """A submission the optimiser thinks you should build.

    Two tiers, because a batch short of its minimum has two answers and the
    difference between them is the whole point of filling it: ``tier_name`` is
    what these cards were routed to and what they would be graded at once the
    batch is full; ``effective_tier_name`` is what they would actually be
    graded at *today*, at the count the batch currently has.
    """

    company_id: str
    company_code: str
    tier_id: str | None
    tier_name: str | None
    #: What the cards actually land on at the current count. Equal to
    #: ``tier_name`` once the batch is viable.
    effective_tier_name: str | None = None
    card_count: int = 0
    minimum_cards: int = 1
    maximum_cards: int | None = None
    short_by: int = 0

    expected_profit: float | None = None
    grading_cost: float | None = None
    #: What this batch would be worth once filled to the tier's minimum, so
    #: "add sixteen more cards" carries a number rather than an instruction.
    expected_profit_if_filled: float | None = None

    viable: bool = True
    reason: str | None = None
    cards: list[PlacedCard] = field(default_factory=list)


@dataclass
class UnplacedCard:
    """A card worth grading that no proposed batch could take."""

    card_id: str
    name: str
    set_label: str | None = None
    company_code: str | None = None
    tier_name: str | None = None
    expected_profit: float | None = None
    reason: str = ""


@dataclass
class OptimiserResult:
    currency: str = "GBP"
    #: Cards with enough behind them to be decided at all.
    analysable: int = 0
    #: Of those, the ones the engine would grade.
    worth_grading: int = 0
    placed: int = 0
    total_cards: int = 0
    truncated: bool = False
    routed_at_batch_size: int = 1

    expected_profit: float | None = None
    total_grading_cost: float | None = None

    batches: list[ProposedBatch] = field(default_factory=list)
    unplaced: list[UnplacedCard] = field(default_factory=list)
    #: Cards that were worth grading until their batch turned out to be small.
    stopped_paying: list[PlacedCard] = field(default_factory=list)

    status: str = BlockStatus.OK.value
    reason: str | None = None
    notes: list[str] = field(default_factory=list)


def _candidates(db: Session) -> list[Card]:
    """Cards with both a current assessment and a computed price."""
    assessed = select(ConditionAssessment.card_id).where(ConditionAssessment.is_current.is_(True))
    priced = select(MarketPrice.catalog_key)
    return list(
        db.scalars(
            select(Card)
            .where(Card.id.in_(assessed), Card.catalog_key.in_(priced))
            .order_by(Card.updated_at.desc())
        )
    )


def hopeful_batch_size(db: Session) -> int:
    """A batch big enough that every tier is on the table.

    Routing at 1 would hide the bulk tiers entirely and route everything to the
    expensive ones; routing at the largest minimum lets each card say where it
    would go if the batch existed, which is the question this pass is asking.
    """
    minimums = db.scalars(
        select(GradingTier.minimum_cards).where(GradingTier.active.is_(True))
    ).all()
    return max([1, *minimums])


def optimise(db: Session, *, limit: int = DEFAULT_LIMIT) -> OptimiserResult:
    """Propose submissions, then check they still pay at the size they came out."""
    settings_values = settings_service.get_all(db)
    result = OptimiserResult(currency=settings_values.get("currency", "GBP"))
    result.total_cards = _count_cards(db)

    candidates = _candidates(db)
    result.analysable = len(candidates)
    if len(candidates) > limit:
        result.truncated = True
        candidates = candidates[:limit]

    routing_size = hopeful_batch_size(db)
    result.routed_at_batch_size = routing_size

    # --- Pass 1: route ------------------------------------------------------
    routed: dict[tuple[str, str | None], list[tuple[Card, object]]] = {}
    for card in candidates:
        evaluated = evaluation.evaluate_card(db, card, batch_size=routing_size)
        recommendation = evaluated.recommendation
        if recommendation.decision not in _WORTH_GRADING:
            continue
        result.worth_grading += 1
        company_code = recommendation.company_code
        if company_code is None:
            result.unplaced.append(
                UnplacedCard(
                    card_id=card.id,
                    name=evaluated.raw.display_name,
                    set_label=evaluated.raw.set_label,
                    expected_profit=recommendation.expected_profit,
                    reason="The engine recommends grading it but named no grader to send it to.",
                )
            )
            continue
        key = (company_code, recommendation.tier_name)
        routed.setdefault(key, []).append((card, evaluated))

    if not routed:
        result.status = BlockStatus.INSUFFICIENT_DATA.value
        result.reason = (
            "Nothing in your collection clears the bar for grading right now, so there is "
            "no submission to build."
            if result.analysable
            else "No card has both a condition assessment and comparable sales yet, so nothing "
            "can be decided — assess a card and add its sales."
        )
        _finish(result, limit)
        return result

    # --- Pass 2 and 3: pack, then re-verify at the real size ----------------
    companies = {company.code: company for company in db.scalars(select(GradingCompany))}
    for (company_code, tier_name), members in sorted(
        routed.items(), key=lambda item: -len(item[1])
    ):
        company = companies.get(company_code)
        if company is None:  # pragma: no cover - routed codes come from the same table
            continue
        tier = _tier_named(company, tier_name)
        batch = ProposedBatch(
            company_id=company.id,
            company_code=company_code,
            tier_id=tier.id if tier else None,
            tier_name=tier_name,
            card_count=len(members),
            minimum_cards=tier.minimum_cards if tier else 1,
            maximum_cards=tier.maximum_cards if tier else None,
        )

        real_size = len(members)
        short = tier is not None and real_size < tier.minimum_cards
        if short:
            batch.short_by = tier.minimum_cards - real_size
            batch.viable = False

        profit_minor = 0
        cost_minor = 0
        for card, routed_eval in members:
            placed = _reverify(db, card, routed_eval, batch, real_size)
            batch.cards.append(placed)
            if placed.still_pays:
                profit_minor += to_minor(placed.expected_profit) or 0
                cost_minor += to_minor(placed.grading_cost) or 0
            else:
                result.stopped_paying.append(placed)

        # Where the cards actually land at this count. They agree with the
        # routed tier once the batch is full, and differ while it is short.
        landed = {card.tier_name for card in batch.cards if card.still_pays}
        batch.effective_tier_name = landed.pop() if len(landed) == 1 else tier_name

        if short:
            filled = sum(
                to_minor(routed_eval.recommendation.expected_profit) or 0
                for _, routed_eval in members
            )
            batch.expected_profit_if_filled = round(filled / 100, 2)
            uplift = format_money(filled - profit_minor)
            same_tier = batch.effective_tier_name == tier_name
            batch.reason = (
                f"{company_code} {tier_name} needs {tier.minimum_cards} cards and this batch has "
                f"{real_size}. "
                + (
                    ""
                    if same_tier
                    else f"As it stands these would be graded at {batch.effective_tier_name}. "
                )
                + f"Adding {batch.short_by} more card(s) at this tier is worth {uplift} across "
                "the cards already in it."
            )

        batch.expected_profit = round(profit_minor / 100, 2)
        batch.grading_cost = round(cost_minor / 100, 2)
        result.batches.append(batch)

    result.batches.sort(key=lambda item: (item.expected_profit or 0), reverse=True)
    result.placed = sum(
        1 for batch in result.batches for card in batch.cards if card.still_pays
    )
    paying = [
        card for batch in result.batches for card in batch.cards if card.still_pays
    ]
    if paying:
        result.expected_profit = round(
            sum(to_minor(card.expected_profit) or 0 for card in paying) / 100, 2
        )
        result.total_grading_cost = round(
            sum(to_minor(card.grading_cost) or 0 for card in paying) / 100, 2
        )

    _finish(result, limit)
    return result


def _count_cards(db: Session) -> int:
    from sqlalchemy import func

    return db.scalar(select(func.count()).select_from(Card)) or 0


def _tier_named(company: GradingCompany, tier_name: str | None) -> GradingTier | None:
    if tier_name is None:
        return None
    for tier in company.tiers:
        if tier.tier_name == tier_name and tier.active:
            return tier
    return None


def _reverify(
    db: Session,
    card: Card,
    routed_eval,
    batch: ProposedBatch,
    real_size: int,
) -> PlacedCard:
    """Cost this card again at the size its batch actually came out at.

    A card routed at twenty-five and placed in a batch of six is a different
    proposition: the postage it was going to share is now split six ways. If
    that flips the decision, the card says so rather than riding along inside a
    total that no longer holds.
    """
    routed_recommendation = routed_eval.recommendation
    placed = PlacedCard(
        card_id=card.id,
        name=routed_eval.raw.display_name,
        set_label=routed_eval.raw.set_label,
        company_code=batch.company_code,
        tier_id=batch.tier_id,
        tier_name=batch.tier_name,
        declared_value=routed_eval.grading_options.declared_value,
        decision_when_routed=routed_recommendation.decision,
        decision_in_batch=routed_recommendation.decision,
        expected_profit=routed_recommendation.expected_profit,
        grading_cost=routed_recommendation.grading_cost,
        opportunity_score=routed_recommendation.opportunity_score,
    )

    if real_size == routed_recommendation.assumed_batch_size:
        _suggest_cheaper_tier(placed, routed_eval)
        return placed

    actual = evaluation.evaluate_card(db, card, batch_size=real_size)
    recommendation = actual.recommendation
    placed.decision_in_batch = recommendation.decision
    placed.opportunity_score = recommendation.opportunity_score

    # The tier the card actually lands on at this count, which is not always the
    # one it was routed to: Bulk needs twenty-five, so in a batch of nine these
    # cards are Economy cards. Reporting them as Bulk — at Economy's price —
    # would quote a route that does not exist at this size.
    placed.tier_name = recommendation.tier_name or batch.tier_name
    placed.tier_id = None if placed.tier_name != batch.tier_name else batch.tier_id

    # The figures for that route at that size. The recommendation itself can be
    # describing a fuller submission (`grade_if_batch_filled`), so prefer the
    # outcome row, which is always costed at the size asked for.
    at_this_size = next(
        (
            outcome
            for outcome in actual.expected_outcomes.outcomes
            if outcome.company_code == batch.company_code
            and outcome.tier_name == placed.tier_name
        ),
        None,
    )
    if at_this_size is not None:
        placed.expected_profit = at_this_size.expected_profit
        placed.grading_cost = at_this_size.grading_cost
    else:
        placed.expected_profit = recommendation.expected_profit
        placed.grading_cost = recommendation.grading_cost

    # `grade_if_batch_filled` was a fair answer while the batch was hypothetical.
    # Here the size is a known fact, so it means "not in the batch you have" —
    # accepting it would let the optimiser bless a plan on the strength of a
    # fuller submission that does not exist.
    if recommendation.decision != Decision.GRADE.value:
        placed.still_pays = False
        was = routed_recommendation.expected_profit
        now = placed.expected_profit
        moved = (
            f" Expected profit falls from {format_money(to_minor(was))} to "
            f"{format_money(to_minor(now))}."
            if was is not None and now is not None and to_minor(was) != to_minor(now)
            else ""
        )
        placed.reason = (
            f"Worth grading in a batch of {routed_recommendation.assumed_batch_size}, but this "
            f"batch has {real_size} — so it carries a bigger share of the postage.{moved}"
        )
    else:
        _suggest_cheaper_tier(placed, actual)
    return placed


def _suggest_cheaper_tier(placed: PlacedCard, evaluated) -> None:
    """"Move it to Bulk and save £2.20" — but only where the move is legal.

    Reads the options block the evaluation already computed, so this is a lookup
    over numbers the user can see rather than a second costing they cannot.
    """
    current = next(
        (
            option
            for option in evaluated.grading_options.options
            if option.company_code == placed.company_code
            and option.tier_name == placed.tier_name
        ),
        None,
    )
    if current is None or current.total_cost is None:
        return

    cheaper = [
        option
        for option in evaluated.grading_options.options
        if option.company_code == placed.company_code
        and option.available
        and option.total_cost is not None
        and option.total_cost < current.total_cost
    ]
    if not cheaper:
        return

    best = min(cheaper, key=lambda option: option.total_cost)
    placed.cheaper_tier_name = best.tier_name
    placed.cheaper_tier_saving = round(current.total_cost - best.total_cost, 2)


def _finish(result: OptimiserResult, limit: int) -> None:
    """Set the status and say anything the numbers above do not."""
    if result.truncated:
        result.notes.append(
            f"Only the {limit} most recently updated ready cards were considered. "
            "Raise the limit to include the rest."
        )
    if result.stopped_paying:
        result.notes.append(
            f"{len(result.stopped_paying)} card(s) were worth grading in a full batch but not "
            "in the batch they ended up in. They are listed against their batch, and are not "
            "counted in any total."
        )
    short = [batch for batch in result.batches if not batch.viable]
    if short:
        result.notes.append(
            f"{len(short)} proposed batch(es) are short of their tier's minimum. They are "
            "costed as they stand, so you can see what filling them would be worth."
        )

    if result.status == BlockStatus.INSUFFICIENT_DATA.value:
        return
    if result.truncated or short or result.stopped_paying or result.unplaced:
        result.status = BlockStatus.PARTIAL.value
        result.reason = result.notes[0] if result.notes else None
    else:
        result.status = BlockStatus.OK.value
